crates/sparrow-merge/src/merge_pipeline.rs

```diff
@@ -0,0 +1,340 @@
+            PipelineError::InputClosed {
+                input,
+                input_partition
+            }
+        );
+
+        if *self.left.is_closed.lock() && *self.right.is_closed.lock() {
+            // Don't close the sink (consumers) here. We may be currently executing a `do_work`
+            // loop, in which case we need to allow it to output to the sink before
+            // we close it.
+            //
+            // We call `close` on the partition to indicate that once the work loop sees
+            // that the partition is empty of inputs, it can close its consumers and
+            // complete itself.
+            partition.close();
+            scheduler.schedule(partition.task.clone());
+        }
+
+        Ok(())
+    }
+
+    fn do_work(
+        &self,
+        input_partition: Partition,
+        scheduler: &mut dyn Scheduler,
+    ) -> error_stack::Result<(), PipelineError> {
+        let partition = &self.partitions[input_partition];
+        let _enter = tracing::trace_span!("MergePipeline::do_work", %partition.task).entered();
+
+        let mut handler = partition.handler.lock();
+        if let Some(active_input) = handler.merger.blocking_input() {
+            let receiver = &mut handler.rxs[active_input];
+            match receiver.try_recv() {
+                Ok(batch) => {
+                    // Add the batch to the active input
+                    let ready_to_produce = handler.merger.add_batch(active_input, batch);
+                    if ready_to_produce {
+                        let merged_batch = handler
+                            .merger
+                            .merge()
+                            .change_context(PipelineError::Execution)?;
+
+                        // If the result is non-empty, output it.
+                        self.consumers
+                            .add_input(input_partition, merged_batch, scheduler)
+                            .change_context(PipelineError::Execution)?;
+                    }
+                }
+                Err(TryRecvError::Empty) => {
+                    error_stack::ensure!(
+                        partition.is_closed(),
+                        PipelineError::illegal_state("scheduled without work")
+                    );
+
+                    // The channel is empty and the partition is closed, but the input may
+                    // not have disconnected from the channel yet. This is okay -- if the partition
+                    // is verified closed, we can continue to close our consumers here.
+                    return self.complete_partition(input_partition, partition, scheduler);
+                }
+                Err(TryRecvError::Disconnected) => {
+                    return self.complete_partition(input_partition, partition, scheduler)
+                }
+            }
+        } else {
+            // Though all inputs are closed, the merger may have batches remaining to flush.
```

Comment:
Is this all closed? This just means "nothing blocking the gatherer", right?

Reply:
I suppose both yeah..we technically know that `close` has been called on all inputs, and we want to assert that that is the case, because we're completing our task here, but the actual `if` is just checking whether the gatherer has any blocking input. I can update language